pkg/slayers/path/scion/info_hop_setter_lemmas.gobra

```diff
@@ -0,0 +1,519 @@
+// Copyright 2022 ETH Zurich
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +gobra
+
+package scion
+
+import (
+	"github.com/scionproto/scion/pkg/slayers/path"
+	. "verification/utils/definitions"
+	sl "verification/utils/slices"
+	"verification/io"
+)
+
+ghost
+decreases
+pure func AbsSetCurrSeg(seg io.IO_seg3, info path.IntermediateAbsInfoField) (io.IO_seg3) {
+	return io.IO_seg3_ {
+		info.AInfo,
+		info.UInfo,
+		info.ConsDir,
+		info.Peer,
+		seg.Past,
+		seg.Future,
+		seg.History,
+	}
+}
+
+ghost
+decreases
+pure func AbsSetInfoField(oldPkt io.IO_pkt2, info path.IntermediateAbsInfoField) (newPkt io.IO_pkt2) {
+	return let newCurrSeg := AbsSetCurrSeg(oldPkt.CurrSeg, info) in
+	io.IO_Packet2{newCurrSeg, oldPkt.LeftSeg, oldPkt.MidSeg, oldPkt.RightSeg}
+}
+
+// SliceBytesIntoSegments splits the raw bytes of a packet into its three hopfield segments
+ghost
+requires 0 < p
+requires 0 < seg1Len
+requires 0 <= seg2Len
+requires 0 <= seg3Len
+requires 0 <= offset
+requires offset + (seg1Len + seg2Len + seg3Len)*path.HopLen <= len(raw)
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), p)
+ensures  acc(sl.AbsSlice_Bytes(raw[:offset], 0, offset), p)
+ensures  acc(sl.AbsSlice_Bytes(raw[offset:offset+seg1Len*path.HopLen], 0, seg1Len*path.HopLen), p)
+ensures  acc(sl.AbsSlice_Bytes(raw[offset+seg1Len*path.HopLen:offset+(seg1Len+seg2Len)*path.HopLen], 0, seg2Len*path.HopLen), p)
```

Comment:
If we have the structure, than we could have functions to compute these offsets without having to write the big ugly operations.